### ISSUE TITLE
snapshot of container with nested lxd with btrfs storage fails

### ISSUE THREAD

**Issue by laralar** [lxc/lxd#4731]
# Required information

 * Distribution: Ubuntu
 * Distribution version: 18.04
 * The output of "lxc info" or if that fails:
   * Kernel version: 4.15.0-23-generic
   * LXC version: snap 3.2
   * LXD version: snap 3.2
   * Storage backend in use: btrfs

# Issue description

I can't create a snapshot of a container that has nested lxd
# Steps to reproduce

```
root@node12:~/scripts/abdf-setup# lxc storage list
+---------+-------------+--------+---------------+---------+
|  NAME   | DESCRIPTION | DRIVER |    SOURCE     | USED BY |
+---------+-------------+--------+---------------+---------+
| data    |             | btrfs  | /data/bpool   | 4       |
+---------+-------------+--------+---------------+---------+
| default |             | zfs    | zpool/default | 31      |
+---------+-------------+--------+---------------+---------+
root@node12:~/scripts/abdf-setup#
root@node12:~/scripts/abdf-setup# lxc storage volume list data
+-----------+------------------------------------------------------------------+-------------+---------+
|   TYPE    |                               NAME                               | DESCRIPTION | USED BY |
+-----------+------------------------------------------------------------------+-------------+---------+
| container | os-client                                                        |             | 1       |
+-----------+------------------------------------------------------------------+-------------+---------+
| container | os-client/aibladmin                                              |             | 1       |
+-----------+------------------------------------------------------------------+-------------+---------+
| container | os-client/lxd-init                                               |             | 1       |
+-----------+------------------------------------------------------------------+-------------+---------+
| image     | b190d5ec0c537468465e7bd122fe127d9f3509e3a09fb699ac33b0c5d4fe050f |             | 1       |
+-----------+------------------------------------------------------------------+-------------+---------+
root@node12:~/scripts/abdf-setup#
```

 1. lxc launch ubuntu:b os-client --storage data
 2. lxc config set os-client security.privileged true
 3. lxc config set os-client security.nesting true
 4. lxc stop os-client
 5. lxc start os-client
 
```
root@node12:~/scripts/abdf-setup# lxc shell os-client
mesg: ttyname failed: No such device
root@os-client:~# cd /var/lib/lxd
root@os-client:/var/lib/lxd# ls
unix.socket
root@os-client:/var/lib/lxd# cd ..
root@os-client:/var/lib# lxd init
Would you like to use LXD clustering? (yes/no) [default=no]:
Do you want to configure a new storage pool? (yes/no) [default=yes]:
Name of the new storage pool [default=default]:
Name of the storage backend to use (btrfs, dir, lvm) [default=btrfs]:
Would you like to create a new btrfs subvolume under /var/lib/lxd? (yes/no) [default=yes]:
Would you like to connect to a MAAS server? (yes/no) [default=no]:
Would you like to create a new network bridge? (yes/no) [default=yes]:
What should the new bridge be called? [default=lxdbr0]:
What IPv4 address should be used? (CIDR subnet notation, “auto” or “none”) [default=auto]: auto
What IPv6 address should be used? (CIDR subnet notation, “auto” or “none”) [default=auto]: none
Would you like LXD to be available over the network? (yes/no) [default=no]:
Would you like stale cached images to be updated automatically? (yes/no) [default=yes]
Would you like a YAML "lxd init" preseed to be printed? (yes/no) [default=no]:
root@os-client:/var/lib# exit
logout
root@node12:~/scripts/abdf-setup# lxc snapshot os-client lxd-init
Error: subvolume snapshot failed, source=/var/snap/lxd/common/lxd/storage-pools/data/containers/os-client/rootfs/var/lib/lxd/storage-pools/default, dest=/var/snap/lxd/common/lxd/storage-pools/data/snapshots/os-client/lxd-init/rootfs/var/lib/lxd/storage-pools/default, output=ERROR: cannot snapshot '/var/snap/lxd/common/lxd/storage-pools/data/containers/os-client/rootfs/var/lib/lxd/storage-pools/default': Inappropriate ioctl for device
Create a snapshot of '/var/snap/lxd/common/lxd/storage-pools/data/containers/os-client/rootfs/var/lib/lxd/storage-pools/default' in '/var/snap/lxd/common/lxd/storage-pools/data/snapshots/os-client/lxd-init/rootfs/var/lib/lxd/storage-pools/default/default'

```
NOTE: the btrfs subvolume was created on the storage pool (see above on storage pool and storage volume information)

Am I missing something?

# Information to attach

 - [ ] Any relevant kernel output (`dmesg`)
 - [ ] Container log (`lxc info NAME --show-log`)
 - [ ] Container configuration (`lxc config show NAME --expanded`)
 - [ ] Main daemon log (at /var/log/lxd/lxd.log or /var/snap/lxd/common/lxd/logs/lxd.log)
 - [ ] Output of the client with --debug
 - [ ] Output of the daemon with --debug (alternatively output of `lxc monitor` while reproducing the issue)

**Comment by stgraber** [comment]
Moving the issue over to the snap packaging https://github.com/lxc/lxd-pkg-snap/issues/19

**Comment by stgraber** [comment]
Actually doesn't appear snap related, just reproduced on my development system. It seems like something broke in our logic to snapshot nested btrfs subvolumes.

**Comment by gschintgen** [comment]
Hi, I think I've hit the same bug while snapshotting a LXD container which is acting as docker host (nested btrfs...). At least it's the exact same error message. I'm using LXD 3.0.2 (via apt) on Ubuntu 18.04.
As far as I can see the fix in #4764 didn't make it into 3.0.2.
Do you think it'd be feasible to have this fix in 3.0.3? Otherwise I'll think about migrating to the snap.
Thanks anyway! It's the first real issue that I hit while running LXD, so I'm quite satisfied with its stability.
